electrotest.c

```diff
@@ -0,0 +1,55 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include "libresistance.h"
+#include "libpower.h"
+#include "libcomponent.h"
+
+void clearInputBuffer() {
+  char c;
+  while((c = getchar()) != '\n' && c != EOF);
+}
+
+int main() {
+  float volt;
+  char conn;
+  int n_comp;
+
+  printf("Ange spänningskälla i V: ");
+  scanf("%f", &volt);
+  clearInputBuffer();
+
+  printf("Ange koppling [P | S]: ");
+  scanf("%c", &conn);
+  clearInputBuffer();
+
+  printf("Antal komponenter: ");
+  scanf("%d", &n_comp);
+  clearInputBuffer();
+
+  // input components
+  float* comp_array = (float*) malloc(n_comp * sizeof(float));
```

Comment:
När man använder malloc måste man alltid kolla så allokeringen lyckades (kolla så att pekaren != NULL)

Reply:
Vad gör man om pekaren inte är NULL? Provar free() igen?

Reply:
Om pekaren är NULL har malloc misslyckats och jag tror man brukar avsluta programmet med ett felmeddelande, om den inte är NULL har allt funkat som det ska och man kan fortsätta som vanligt. Så egentligen e det bara att lägga till typ:
if(comp_array == NULL) {
fprintf(stderr, "Failed to allocate memory\n");
exit(1);
}
eller nåt sånt.
Ett annat alternativ är att definiera en ny malloc typ:
void* xmalloc(size_t size) {
void* mem = malloc(size);
if(mem == NULL) {
fprintf(stderr, "Failed to allocate memory\n");
exit(1);
}
return mem;
}

Reply:
Förmodligen sant i en större applikation (använder själv i princip alltid språk där minnesallokering är bortabstraherat), men är det verkligen någonting vi behöver bry oss om för den här lilla labben? T.ex. brukar jag aldrig bry mig om felhantering i labbar om det inte explicit står som krav, även om detta givetvis bör finnas i "riktiga" applikationer.

Reply:
Nä, förmodligen inte :-)

Reply:
Mm, jag hade inte använt det om det inte stått specifikt i uppgiften:

> Använd dynamisk minnesallokering med någon variant av malloc och free.